components/num-util/src/fixed_decimal.rs

```diff
@@ -415,94 +420,123 @@ impl FromStr for FixedDecimal {
+            }
+            if c == b'0' {
                 right_zeros += 1;
             } else {
+                rightmost_digit = no_sign_str_len - i;
                 break;
             }
         }
-        // digits_str: the string after removing sign, dot, left zeros, and right zeros
-        let digits_str = &no_dot_str[left_zeros..no_dot_str_len - right_zeros];
 
-        // shift: number of digits after dot in the input_str, 0 if there is no dot
-        let mut shift: i16 = 0;
-        if has_dot {
-            shift = (no_sign_str_len - dot_index - 1) as i16;
+        // let digits_str_len: shows the length of digits (Ex. 0012.8900 --> 4)
+        let mut digits_str_len = rightmost_digit - leftmost_digit;
+        if leftmost_digit < dot_index && dot_index < rightmost_digit {
+            digits_str_len -= 1;
         }
 
-        // Constructing DecimalFixed only using digits_str
-        let mut dec: FixedDecimal = 0.into();
-        let mut v = SmallVec::<[u8; 8]>::new();
-        for (i, c) in digits_str.chars().enumerate() {
+        // Constructing DecimalFixed.digits
+        let mut v: SmallVec<[u8; 8]> = SmallVec::with_capacity(digits_str_len);
+        for (i, c) in no_sign_str[leftmost_digit..rightmost_digit]
+            .bytes()
+            .enumerate()
+        {
+            if c == b'.' {
+                continue;
+            }
             if i > std::i16::MAX as usize {
                 return Err(Error::Limit);
             }
-            // ord('0') = 48
-            v.push(c as u8 - 48);
+            v.push(c - b'0');
         }
         let v_len = v.len();
+        debug_assert_eq!(v_len, digits_str_len);
+        debug_assert_eq!(
+            no_sign_str_len,
+            left_zeros + right_zeros + v_len + has_dot as usize
+        );
         dec.digits = v;
-        dec.magnitude = v_len as i16 - 1;
-        dec.upper_magnitude = v_len as i16 - 1;
-        dec.lower_magnitude = 0;
-
-        // Adjusting magnitude, upper_magnitude, and lower_magnitud w.r.t. left_zeros, right_zeros, and shift
-        // The computed upper_magnitude value computed below is always >= 0
-        // The computed lower_magnitude value computed below is always <= 0
-        // The order of negative terms and positive terms is intentional, to avoid overflow if possible
-        dec.upper_magnitude += -shift + left_zeros as i16 + right_zeros as i16;
-        dec.magnitude += -shift + right_zeros as i16;
-        dec.lower_magnitude -= shift;
-        dec.is_negative = is_negative;
 
-        // Outputs for debugging:
-        // println!("digits_str = {}", digits_str);
-        // println!("dot_index = {}", dot_index);
-        // println!("right zeros = {}", right_zeros);
-        // println!("left zeros = {}", left_zeros);
-        // println!("shift = {}", shift);
-        // println!("final_dec = {}", dec);
-        // println!("final_dec = {:?}", dec);
+        // Computing DecimalFixed.upper_magnitude
```

Comment:
Since this code is exactly the same as in the all-zeros case, I suggest moving this up in the function so that it can be shared.